persistence/sql/persister_session.go

```diff
@@ -82,15 +83,34 @@ func (p *Persister) UpsertSession(ctx context.Context, s *session.Session) error
 
 	s.NID = p.NetworkID(ctx)
 
-	if err := p.Connection(ctx).Find(new(session.Session), s.ID); errors.Is(err, sql.ErrNoRows) {
-		// This must not be eager or identities will be created / updated
-		return errors.WithStack(p.GetConnection(ctx).Create(s))
-	} else if err != nil {
-		return errors.WithStack(err)
-	}
+	return errors.WithStack(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) error {
+		if err := tx.Find(new(session.Session), s.ID); errors.Is(err, sql.ErrNoRows) {
+			// This must not be eager or identities will be created / updated
+			err = errors.WithStack(tx.Create(s))
+			if err != nil {
+				return err
+			}
+
+			for i := range s.Devices {
+				device := &(s.Devices[i])
+				device.SessionID = s.ID
+				device.NID = s.NID
 
-	// This must not be eager or identities will be created / updated
-	return p.GetConnection(ctx).Update(s)
+				if err := tx.Create(device); err != nil {
+					return err
+				}
+
+				s.Devices[i] = *device
+			}
+
+			return nil
+		} else if err != nil {
+			return errors.WithStack(err)
+		}
+
+		// This must not be eager or identities will be created / updated
+		return tx.Update(s)
```

Comment:
So this won't update devices if I read this correctly. If that's the case, plase document it in the code doc comment(s) of the method / interface and here!

Reply:
Added notes to the func